### ISSUE TITLE
Using `notFound` in suspended interleaved server component breaks rendering if not-found page (or layout) renders a `next/link` - production-only - Minified React error #310

### ISSUE THREAD

**Issue by michaelschufi** [vercel/next.js#63388]
### Link to the code that reproduces this issue

https://codesandbox.io/p/devbox/elastic-sound-fpmpdt

### To Reproduce

1. Build and run the project with `pnpm build && pnpm start` if it's not already running.
1. Open https://fpmpdt-3000.csb.app/en
3. If no client-side exception shows up, hard-refresh until it does.
4. Observe the `Minified React error #310` (Rendered more hooks than during the previous render.) being logged in the browser's console.

Note:
The project has to be built. In dev-mode, no error occurs. It's also not always the case.

### Current vs. Expected behavior

I expect the not-found page inside `/[locale]` to be rendered without issue. However, the page rendering results in a React error.

### Provide environment information

_Original_
```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #1 SMP PREEMPT_DYNAMIC Sun Aug  6 20:05:33 UTC 2023
  Available memory (MB): 4102
  Available CPU cores: 2
Binaries:
  Node: 20.9.0
  npm: 9.8.1
  Yarn: 1.22.19
  pnpm: 8.10.2
Relevant Packages:
  next: 14.2.0-canary.26 // Latest available version is detected (14.2.0-canary.26).
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.1.3
Next.js Config:
  output: N/A
```
Also tested with
```
Operating System:
  Platform: linux
  Arch: x64
  Version: #1 SMP Thu Jan 11 04:09:03 UTC 2024
  Available memory (MB): 15853
  Available CPU cores: 16
Binaries:
  Node: 18.20.1
  npm: 10.5.0
  Yarn: 1.22.22
  pnpm: 8.15.5
Relevant Packages:
  next: 14.3.0-canary.73 // Latest available version is detected (14.3.0-canary.73).
  eslint-config-next: N/A
  react: 19.0.0-beta-04b058868c-20240508
  react-dom: 19.0.0-beta-04b058868c-20240508
  typescript: 5.1.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router, Routing (next/router, next/navigation, next/link)

### Which stage(s) are affected? (Select all that apply)

next start (local), Vercel (Deployed)

### Additional context

First of all, the crash won't happen 100% of the time, and only in production mode. This took a while to debug 😅

**The stars need to align**
The crash stops happening, or happens less often, if one of the following is the case:
- The `[locale]/not-found.tsx` page does not include a `next/link` `<Link>`.
- The `<Suspense>` is removed.
- The `useRouter` and its `useEffect` is commented out.
- The data fetching in the server component is taking longer. E.g. `500ms` instead of the `50ms` configured. This points to a race condition.

**Race condition**
Because of the race-conditiony nature of the bug, it could not be reproducible on some machines.

**`next/link` links in general?**
It seems that any next/link link that will be rendered together with the not found page (e.g. those in the layout) triggers the bug. In a bigger project (where the issue first appeared), turning all the `<Link>` tags into `<a>` tags got rid of the issue.

**Comment by slackerzz** [comment]
Like @andremendonca, I use page streaming and I get this random error when I call `notFound()`

Disabling streaming (deleting `loading.tsx` file) fixes the issue

**Comment by andremendonca** [comment]
> Like @andremendonca, I use page streaming and I get this random error when I call `notFound()`
> 
> Disabling streaming (deleting `loading.tsx` file) fixes the issue

Exactly what I had to do to "fix" the issue.

**Comment by michaelschufi** [comment]
Yes, but that's the point. The suspend boundary is needed for our case (better UX).

**Comment by slackerzz** [comment]
> Yes, but that's the point. The suspend boundary is needed for our case (better UX).

It's a workaround of course. Having no errors is clearly a UX improvement 😅

**Comment by michaelschufi** [comment]
It's not a workaround. You won't get similar behaviour at all. 

Not using a feature doesn't work around a bug of said feature.
Imagine you open an issue because SCSS doesn't work with a framework. "Just don't use SCSS." is not a workaround. Maybe "compile SCSS in a build-step instead of on-demand" could be.

**Comment by SaveliiLukash** [comment]
Facing a similar issue using `next@14.2.2`. We have a `page.tsx` with `not-found.tsx` and `loading.tsx` files.

Once the `page.tsx` calls the `notFound()` function the `not-found.tsx` component is displayed with either **Minified React error <span>#</span>419** (the app remains alive) or **Minified React error <span>#</span>310** (which crashes the app).

Happens in production only, not reproducible in development.

As noted here by kind people removing `loading.tsx` instantly fixed all Minified React errors. This, at least, keeps our app stable, but, of course worsens the UX by not providing instant navigation with a loading state.

**Comment by michaelschufi** [comment]
@SaveliiLukash What happens if you remove all `next/link` usages? Does it work then?

**Comment by josh-feldman** [comment]
I'm having the same issue using next@14.1.4, but we are not using a not-found page or next/link on the layout. We have a next/link on the loading.tsx page, and then we are using a `redirect`, NOT a `notFound` and getting the same issue

**Comment by Zakariaels** [comment]
Hello,

Is there any news about this bug ?

**Comment by FengeRVictor** [comment]
Same issue for me, when I'm removing the loading.tsx, it works well. 
On my page, I have a redirect to a page that is using a loading.tsx

**Comment by websaid** [comment]
Having the same issue. Please fix if possible.

**Comment by zjnka141** [comment]
Same issue for me. But I'm not using the loading.tsx file, I use <Suspend> with a skeleton loader to wrap the page. When I remove the Suspend, it works well.
Happens on Nextjs 14.2.3.

**Comment by yarikpetrenko** [comment]
+1

**Comment by tristan-ieee** [comment]
Same issue here, removing the loading.tsx file seems to fix the issue. 
This is a UX issue for us- will this be addressed in an upcoming release?

**Comment by secure12** [comment]
We are facing the same issue as well. https://stackoverflow.com/questions/78945085/next-js-rendered-more-hooks-than-during-the-previous-render-when-notfound-is-use

In our example, there is a provider in which there is an effect which calls a server action. In the provider, there is a React Suspense wrapping an async component, which can call `notFound`.

**Comment by paigekim29** [comment]
Having the same issue on v14.2.1 using redirect

**Comment by ext0** [comment]
Running into this error as well. Removing the loading.tsx on the sub-page doesn't seem to work.

**Comment by JKostov** [comment]
We have this problem because we are using the `next-intl` package.
Versions of the packages are:`"next": "14.2.3"` and `"next-intl": "^3.15.2"`.

We had to create `layout.tsx` outside `[lang]` and here we also put the `loading.tsx`. That way it works.

❌ Just to be clear with this folder structure **it doesn't work**:
```
└── app
    └── [lang]
          ├── layout.tsx
          ├── loading.tsx
          ├── not-found.tsx
          ├── error.tsx
          └── page.tsx
```

✅ With this it **works**:

```
└── app
    ├─── loading.tsx
    ├─── layout.tsx
    ├─── global-error.tsx
    └──[lang]
        ├── layout.tsx
        ├── not-found.tsx
        ├── error.tsx
        └── page.tsx
```

**Comment by Stefan94V** [comment]
It took a day of my life, but disabling loading.tsx solved this weard behaviour..

**Comment by AyronK** [comment]
This is still an issue in Next 15.1.6

**Comment by cpotey** [comment]
Just to chime in, also am replicating this in `"next": "14.2.25"`.

Such an odd one. Will have to remove the `loading.tsx` files for now, until a better solution appears.

**Comment by klonwar** [comment]
Same for me, was very surprised that such basic functionality cannot work together.

**Comment by ztanner** [comment]
Hi all -- we believe this is a React bug, which we've filed a minimal repro for [here](https://github.com/facebook/react/issues/33580), and we'll keep an eye on the progress so that we can sync the fix into Next.js once it lands.

In the meantime, I will close this in favor of having a single report for this issue. Thank you for providing us with a reproduction 🙏

**Comment by gaearon** [comment]
https://github.com/vercel/next.js/issues/63121#issuecomment-4847557319
